### DISCUSSION TITLE
How can I create a dynamic App?

### DISCUSSION THREAD

**Discussion by cam510** [actix/actix-web#2821] [Q&A] [answered]
Hi, I want to create a dynamic App but have a complime error below:

type mismatch resolving `<impl ServiceFactory<ServiceRequest, Response = ServiceResponse<EitherBody<EitherBody<Encoder<BoxBody>>>>, Error = actix_web::Error, Config = (), InitError = ()> as ServiceFactory<ServiceRequest>>::Response == ServiceResponse<B>`
expected struct `ServiceResponse<EitherBody<EitherBody<Encoder<BoxBody>>>>`
   found struct `ServiceResponse<B>`rustc[E0271](https://doc.rust-lang.org/error-index.html#E0271)
main.rs(181, 13): this type parameter

I enter the warp function, it's also return the 
```rust
App<
    impl ServiceFactory<
        ServiceRequest,
        Response = ServiceResponse<B>,
        Error = actix_web::Error,
        Config = (),
        InitError = (),
    >,
>
```

My code below:
```rust
fn init_app<B>() -> App<
    impl ServiceFactory<
        ServiceRequest,
        Response = ServiceResponse<B>,
        Error = actix_web::Error,
        Config = (),
        InitError = (),
    >,
> where B: MessageBody
{
    let init_db_url = PROJECT_CONFIGS.db_url.as_ref();
    let pg_conn = move || PgConnection::connect2(init_db_url);
    let debug_app = App::new()
        .data_factory(pg_conn)
        .wrap(middleware::Compress::default())
        .wrap(Cors::default().allow_any_origin());
    let release_app = App::new()
        .data_factory(pg_conn)
        .wrap(middleware::Compress::default());
    #[cfg(debug_assertions)]
    if cfg!(debug_assertions) {
        debug_app
    } else {
        release_app
    }
}
```

Thanks :)

**Comment by robjtede** [comment, marked as answer]
Dynamic apps are difficult for the same reason as dynamic anything is difficult in Rust.

The workaround for dynamic middleware is the [`Condition`](https://docs.rs/actix-web/latest/actix_web/middleware/struct.Condition.html) middleware wrapper.

**Reply by cam510** [reply]
Thanks ^_^

**Comment by Samielakkad** [comment]
The problem is that every `.wrap(...)` changes the concrete `App` type. In Rust, `if`/`else` branches must return the same concrete type, but these two apps do not have the same type once one has CORS and the other does not.

For conditional middleware, use `middleware::Condition`:

```rust
use actix_web::middleware::Condition;

let cors_enabled = PROJECT_CONFIGS.cors_enabled;

App::new()
    .data_factory(pg_conn)
    .wrap(middleware::Compress::default())
    .wrap(Condition::new(
        cors_enabled,
        Cors::default().allow_any_origin(),
    ))
```

That keeps the outer app type stable while enabling/disabling the middleware at runtime/config time.

If the app structure itself is truly dynamic, another option is to move the conditional logic inside services/handlers instead of trying to return different `App` builder types. But for middleware on/off switches, `Condition` is the intended tool.

Also avoid trying to name the full `App<impl ServiceFactory<...>>` return type unless you really need to. Actix app builder types get complicated very fast; usually it is easier to build the `App` directly inside the `HttpServer::new` closure.
